dropshot/src/server.rs

```diff
@@ -545,18 +566,43 @@ type InnerHttpsServerStarterNewReturn<C> =
 impl<C: ServerContext> InnerHttpsServerStarter<C> {
     /// Begins execution of the underlying Http server.
     fn start(
-        self,
-        close_signal: tokio::sync::oneshot::Receiver<()>,
+        mut self,
+        mut close_signal: tokio::sync::oneshot::Receiver<()>,
         log_close: Logger,
-    ) -> tokio::task::JoinHandle<Result<(), hyper::Error>> {
-        let graceful = self.0.with_graceful_shutdown(async move {
-            close_signal.await.expect(
-                "dropshot server shutting down without invoking close()",
-            );
-            info!(log_close, "received request to begin graceful shutdown");
-        });
+    ) -> tokio::task::JoinHandle<()> {
+        use hyper_util::rt::{TokioExecutor, TokioIo, TokioTimer};
+        use hyper_util::server::conn::auto;
+
+        tokio::spawn(async move {
+            let mut builder = auto::Builder::new(TokioExecutor::new());
+            // http/1 settings
+            builder.http1().timer(TokioTimer::new());
+
+            let graceful =
+                hyper_util::server::graceful::GracefulShutdown::new();
+
+            loop {
+                tokio::select! {
+                    Some(Ok(sock)) = self.0.accept() => {
+                        let remote_addr = sock.remote_addr();
+                        let fut = builder.serve_connection_with_upgrades(
+                            TokioIo::new(sock),
+                            self.1.make_svc(remote_addr),
+                        );
+                        let fut = graceful.watch(fut.into_owned());
+                        tokio::spawn(fut);
+                    },
+
+                    _ = &mut close_signal => {
+                        info!(log_close, "received request to begin graceful shutdown");
+                        break;
+                    }
+                }
+            }
 
-        tokio::spawn(graceful)
+            // optional: could use another select on a timeout
+            graceful.shutdown().await
```

Comment:
i.e. to return early even if shutdown() didn't complete? What would the state of the system be in that case?

Reply:
It would then be an abrupt shutdown. You might have some connections that still have requests/responses in-flight, but streaming for forever. In many systems, you can't wait for forever. But if you have something that will kill the process completely, then perhaps there's less value.

Reply:
To confirm -- is this the same behavior as the old code -- i.e. did the old code not have a timeout on shutdown?

Can this code be combined with `InnerHttpServerStarter` above? Seems like it's complex enough (especially if we add a second select loop for this) that it's worth factoring out.

Reply:
The current code in the PR is the same behavior as before. There was not a timeout before. I can remove the "optional" comment if it's confusing, I meant it as a possible addition down the line.

Reply:
It's fine to leave it for future generations to ponder.